python/lsst/sims/catUtils/utils/SNIaLightCurveGenerator.py

```diff
@@ -0,0 +1,164 @@
+from lsst.sims.catUtils.utils import _baseLightCurveCatalog
+from lsst.sims.catUtils.utils import LightCurveGenerator
+
+from lsst.sims.catUtils.supernovae import SNObject, SNUniverse
+from lsst.sims.photUtils import PhotometricParameters, calcGamma
+from lsst.sims.photUtils import Sed, calcMagError_m5, BandpassDict
+
+import time
+
+__all__ = ["SNIaLightCurveGenerator"]
+
+
+class _sniaLightCurveCatalog(_baseLightCurveCatalog, SNIaCatalog, PhotometryBase):
+
+    column_outputs = ["uniqueId", "snid", "raJ2000", "decJ2000",
+                      "cosmologicalDistanceModulus", "redshift", "EBV"]
+
+    _suppressDimSN = False
+
+
+class SNIaLightCurveGenerator(LightCurveGenerator):
+
+    def __init__(self, *args, **kwargs):
+        self.lsstBandpassDict = BandpassDict.loadTotalBandpassesFromFiles()
+        self._lightCurveCatalogClass = _sniaLightCurveCatalog
+        self._filter_cat = None
+        self._ax_cache = None
+        self._bx_cache = None
+        self._ax_bx_wavelen = None
+        self.phot_params = PhotometricParameters()
+        self.sn_universe = SNUniverse()
+        self.sn_universe.suppressDimSN = False
+        self.z_cutoff = 1.2
+        super(SNIaLightCurveGenerator, self).__init__(*args, **kwargs)
+
+    class _filterCatalogClass(_sniaLightCurveCatalog):
+        column_outputs = ["uniqueId", "t0"]
+
+    def _light_curves_from_query(self, cat_dict, query_result, grp):
+
+        t_dict = {}
+        gamma_dict = {}
+        m5_dict = {}
+        t_min = None
+        t_max = None
+        for bp_name in cat_dict:
+            self.lsstBandpassDict[bp_name].sbTophi()
+            t_dict[bp_name] = np.array([obs.mjd.TAI for obs in grp
+                                        if obs.bandpass == bp_name])
+
+            m5_dict[bp_name] = np.array([obs.m5[bp_name] for obs in grp
+                                         if obs.bandpass == bp_name])
+
+            gamma_dict[bp_name] = np.array([calcGamma(self.lsstBandpassDict[bp_name],
+                                                      obs.m5[obs.bandpass], self.phot_params)
+                                            for obs in grp if obs.bandpass == bp_name])
+
+            if len(t_dict[bp_name]) > 0:
+                local_t_min = t_dict[bp_name].min()
+                local_t_max = t_dict[bp_name].max()
+                if t_min is None or local_t_min < t_min:
+                    t_min = local_t_min
+
+                if t_max is None or local_t_max > t_max:
+                    t_max = local_t_max
+
+        snobj = SNObject()
+
+        cat = cat_dict[cat_dict.keys()[0]]  # does not need to be associated with a bandpass
+
+        for chunk in query_result:
+            t_start_chunk = time.time()
+            for sn in cat.iter_catalog(query_cache=[chunk]):
+                sn_rng = self.sn_universe.getSN_rng(sn[1])
+                sn_t0 = self.sn_universe.drawFromT0Dist(sn_rng)
+                if sn[5] <= self.z_cutoff and np.isfinite(sn_t0) and \
```

Comment:
It looks like suppressHighzSN will set z=np.nan if suppressDimSN is set to True.  However, it is not clear to me that, if z=np.nan, drawSNParams (and, more importantly, drawFromX0Dist, since that is what spends the most time) is not called (i.e. even if we suppress the high z supernovae, I *think* sncat.py still spends time calculating x0 values for the supernovae, which is what I am trying to avoid).